modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_remote_downsampler_temporary_storage.py

```diff
@@ -0,0 +1,47 @@
+import numpy as np
+from modyn.trainer_server.internal.trainer.remote_downsamplers.abstract_remote_downsample_strategy import (
+    AbstractRemoteDownsamplingStrategy,
+)
+from modyn.trainer_server.internal.trainer.remote_downsamplers.sample_then_batch_temporary_storage import (
+    SampleThenBatchTemporaryStorage,
+)
+
+
+class AbstractRemoteDownsamplingTemporaryStorage(AbstractRemoteDownsamplingStrategy):
+    def __init__(self, pipeline_id: int, trigger_id: int, batch_size: int, params_from_selector: dict) -> None:
+        super().__init__(pipeline_id, trigger_id, batch_size, params_from_selector)
+
+        if self.sample_before_batch:
+            assert "maximum_keys_in_memory" in params_from_selector
+            self.sample_then_batch_temporary_storage = SampleThenBatchTemporaryStorage(
+                self.pipeline_id,
+                self.trigger_id,
+                self.batch_size,
+                self.downsampled_batch_ratio,
+                params_from_selector["maximum_keys_in_memory"],
+            )
+            self._next_file_idx = 0
+            self._total_files = 0
+
+    def setup_sample_then_batch(self) -> None:
+        assert self.sample_before_batch
+        self._sampling_concluded = False
+        self.sample_then_batch_temporary_storage.reset_temporary_storage()
+
+    def end_sample_then_batch(self) -> None:
+        assert self.sample_before_batch
+        self._sampling_concluded = True
+        self.sample_then_batch_temporary_storage.end_accumulation()
+        self._next_file_idx = 0
+        self._total_files = self.sample_then_batch_temporary_storage.get_total_number_of_files()
+
+    def samples_available(self) -> bool:
+        assert self._sampling_concluded
+        assert self.sample_before_batch
+        return self._next_file_idx < self._total_files
+
+    def get_samples(self) -> np.ndarray:
```

Comment:
ok so this goes a bit together with my comment on the if/else for the params: there are two options:

a) we manage to unify the downsampler interface for batch-then-sample/sample-then-batch, i.e., the downsampler is not aware of the mode it is used in and sample-then-batch/batch-then-sample is just a question of how we use the downsampler, not the downsampler itself (best solution I think)

b) otherwise we should think about whether we can give this some more structure. e.g., by having a StBDownsampler and a BtSDownsampler that share a common Downsampler supercalls. E.g. StBLossDownsampler, BtSLossDownsampler, which both inherit from LossDownsampler. I basically want to reduce the number of if/else branches and have clear responsibilities of the objects. what do you think?